pubtools/_pulp/tasks/delete.py

```diff
@@ -0,0 +1,735 @@
+    """Remove specific content from one or more Pulp repository"""
+
+    def __init__(self, *args, **kwargs):
+        super(Delete, self).__init__(*args, **kwargs)
+        self.to_await = []
+
+    def add_args(self):
+        super(Delete, self).add_args()
+
+        self.add_publisher_args(self.parser)
+
+        self.parser.add_argument(
+            "--repo",
+            help="remove content from these comma-seperated repositories ",
+            type=str,
+            action=SplitAndExtend,
+            split_on=",",
+        )
+
+        self.parser.add_argument(
+            "--advisory",
+            help="remove packages and modules in these comma-separated advisories",
+            type=str,
+            action=SplitAndExtend,
+            split_on=",",
+        )
+
+        self.parser.add_argument(
+            "--file",
+            help="remove these comma-separated rpm, srpm, modulemd or iso file(s)",
+            type=str,
+            action=SplitAndExtend,
+            split_on=",",
+        )
+
+        self.parser.add_argument(
+            "--signing-key",
+            help="remove the content with these signing-keys(s)",
+            type=str,
+            action=SplitAndExtend,
+            split_on=",",
+        )
+
+        self.parser.add_argument(
+            "--no-publish",
+            help="skip publishing the repos and cleaning the cache",
+            action="store_true",
+        )
+
+    def fail(self, *args, **kwargs):
+        LOG.error(*args, **kwargs)
+        sys.exit(30)
+
+    def run(self):
+        repos = {}
+        file_fs = []
+        signing_keys = self.args.signing_key
+        repo_names = self.args.repo
+        advisory = self.args.advisory
+
+        if not (self.args.file or self.args.advisory):
+            self.fail("One of --file or --advisory is required")
+
+        if not self.args.repo:
+            self.fail("Repository names in --repo is required")
+
+        # separates files into rpms, isos and modules
+        rpms, files, modules = self.separate_files(self.args.file or [])
+
+        # delete files
+        file_f = self._delete_standalone_files(
+            repo_names, signing_keys, rpms, files, modules
+        )
+
+        # delete packages from advisory
+        if advisory:
+            file_fs = self.delete_from_advisory(advisory, signing_keys, repo_names)
+
+        # wait for the futures to resolve
+        # file_f returns a list of futures for content removal on each repo
+        # file_fs is a list of file_f from deletion of set of rpms and modules in
+        # each advisory
+        file_fs.append(file_f)
+        fs = f_sequence(file_fs)
+        cleared_repos = [cr for sublist in fs.result() for cr in sublist]
+        for cr in cleared_repos:
+            repo = cr.repo
+            repos.setdefault(repo.id, repo)
+
+        # publish affected repos
+        if repos and not self.args.no_publish:
```

Comment:
This should probably be divided into steps.

- This part should go into a `@step("Publish")`, that would also help you get `--skip publish` working.
- Most of the steps above should probably go into a `@step("Delete content")` or similar.

Reply:
Added steps for specific content type like ```@step("Delete advisories")``` or ```@step(Delete RPMs"")``` instead of general ```@step("Delete content")``` as it provided more granular control on the steps.